### ISSUE TITLE
Support for OAuth

### ISSUE THREAD

**Issue by bsilverm** [node-red/node-red#719]
A number of folks have been requesting OAuth support through the forum.  The HTTP node currently has support for basic auth but there is no solution for OAuth without developing a node (ie - the Twitter node).  I hope this is the right place to be opening this request.

**Comment by knolleary** [comment]
To be clear, do you mean the http request node (using oauth to make requests of other services) or to be able to secure

**Comment by knolleary** [comment]
Oops - that'll teach me to reply to issues with baby on my lap - she'll hit any button she sees.

Was going to say.... Or to secure HTTP in nodes with oauth?

There are reasons why we don't yet support oauth in either of those cases, something I'm sure I've explained on the mailing list in response to those couple of requests.

**Comment by bsilverm** [comment]
Thanks for the reply.  Sounds like I should join the mailing list.  This request is around using the HTTP request node to make requests to OAuth services (rather than securing HTTP in nodes using OAuth).  

For example, I wanted to be able to issue Twitter API requests from within NodeRED, but ended up doing this instead (see below) given it would be difficult to do things like generate an OAuth signature/nonce/etc from within NodeRED.

http://flows.nodered.org/flow/162cd122fcdeb2c0f1c6

**Comment by cowchimp** [comment]
@knolleary is it possible to reconsider this now that the HTTP request node uses the popular `request` library under-the-covers in #1736?

the `request` library supports OAuth out-of-the-box
https://github.com/request/request#oauth-signing

If we want to keep it minimal we don't have to expose it via the HTTP request dialog, but only configuring it using the incoming `msg`. If `msg` includes a `oauth` property, simply copy it to the `opts` that we pass to `request`.

Would be happy to submit a PR.

**Comment by knolleary** [comment]
@cowchimp yes - the move to the `request` module was a prerequisite for this and a couple other feature requests on the node.

I think your proposal of passing `msg.oauth` straight through is sensible at first glance, but I suspect we need to do something slightly different. There is a risk it will break an existing flow that passes in that property for unrelated reasons. 

Another request we've had is to be able to passing username/password credentials for Basic auth - so they don't have to be hardcoded into the node. Currently users can only do that by hand crafting the `Authentication` http header and passing that in.

Given user/password and oath all fall into the same  category, I wonder if there should be a single message property that can be used to pass either/or/both of those things in - with a name that minimises the chance of breaking existing flows (although that risk always exists when nodes start responding to new msg properties).

How about:
```
msg.requestAuth = {
   username: 'fred',
   password: 'sekret',
   oauth: { ... }
}
```

**Comment by knolleary** [comment]
Closing as part of a tidy-up of the issue tracker — this has been inactive for a long time and much has changed in Node-RED since. If it's still relevant on a current version, please open a fresh issue with up-to-date details and we'll take a look. Thanks for raising it.
